sotopia/experimental/agents/moderator.py

```diff
@@ -68,43 +68,143 @@ def __init__(
-            )
+            if output_channel in action.observations_map:
+                await self.r.publish(
+                    output_channel,
+                    Message[output_channel_type](  # type:ignore[valid-type]
+                        data=action.observations_map[output_channel]
+                    ).model_dump_json(),
+                )
 
     async def __aenter__(self) -> Self:
         print(self.scenario)
-        await self.send(
-            Observations(
+        asyncio.create_task(self.booting())
+        self.task_scheduler = asyncio.create_task(self._task_scheduler())
+        return await super().__aenter__()
+
+    async def _task_scheduler(self) -> None:
+        await self.all_agents_awake.wait()
+        return await super()._task_scheduler()
+
+    async def booting(self) -> None:
+        """
+        1. send checking message to agents for every 5 seconds, until all agents are awake
+        - the agent should use this waking message to send their information to moderator for record
+        - if further information of the agents are needed, should be communicated through the booting process
+        2. after all agents are awake, send agent[0] a message to allow the agent to start speaking
+        """
+        while not self.all_agents_awake.is_set():
+            await self.send(
+                Observations(
+                    observations_map={
+                        output_channel: Observation(
+                            agent_name="moderator",
+                            last_turn=self.scenario,
+                            turn_number=0,
+                            available_actions=["none"],
+                        )
+                        for output_channel, agent_name in self.agent_mapping.items()
+                    }
+                )
+            )
+            await asyncio.sleep(5)
+            while not self.observation_queue.empty():
+                agent_action = await self.observation_queue.get()
+                self.agents_awake[agent_action.agent_name] = True
+                self.agent_models[agent_action.agent_name] = agent_action.argument
+            if False not in self.agents_awake.values():
+                self.all_agents_awake.set()
+
+        for output_channel, agent_name in self.agent_mapping.items():
+            if agent_name == self.agents[0]:
+                await self.send(
+                    Observations(
+                        observations_map={
+                            output_channel: Observation(
+                                agent_name="moderator",
+                                last_turn=self.scenario,
+                                turn_number=0,
+                                available_actions=self.available_actions,
+                            )
+                        }
+                    )
+                )
+                break
+        self.current_agent_index += 1
+
+    async def save(self) -> None:
```

Comment:
maybe have a hyperparameter about whether push to DB? and return epilog instead?